rustworkx-core/src/coloring.rs

```diff
@@ -13,13 +13,90 @@
 use std::cmp::Reverse;
 use std::hash::Hash;
 
+use crate::connectivity::isolates;
 use crate::dictmap::*;
 use crate::line_graph::line_graph;
 use hashbrown::{HashMap, HashSet};
 use petgraph::graph::NodeIndex;
-use petgraph::visit::{EdgeCount, EdgeRef, IntoEdges, IntoNodeIdentifiers, NodeCount};
+use petgraph::visit::{
+    EdgeCount, EdgeRef, GraphBase, GraphProp, IntoEdges, IntoNeighborsDirected,
+    IntoNodeIdentifiers, NodeCount, NodeIndexable,
+};
+use petgraph::{Incoming, Outgoing};
 use rayon::prelude::*;
 
+/// Compute a two-coloring of a graph
+///
+/// If a two coloring is not possible for the input graph (meaning it is not
+/// bipartite), `None` is returned.
+///
+/// Arguments:
+///
+/// * `graph` - The graph to find the coloring for
+///
+/// # Example
+///
+/// ```rust
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::coloring::two_color;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = UnGraph::<i32, i32>::from_edges(&edge_list);
+/// let coloring = two_color(&graph).unwrap();
+/// let mut expected_colors = DictMap::new();
+/// expected_colors.insert(NodeIndex::new(0), 1);
+/// expected_colors.insert(NodeIndex::new(1), 0);
+/// expected_colors.insert(NodeIndex::new(2), 1);
+/// expected_colors.insert(NodeIndex::new(3), 0);
+/// expected_colors.insert(NodeIndex::new(4), 1);
+/// assert_eq!(coloring, expected_colors)
+/// ```
+pub fn two_color<G>(graph: G) -> Option<DictMap<G::NodeId, u8>>
```

Comment:
A general question about Rust: is there a performance penalty from using `DictMap` vs a vector (indexed by NodeIndex or something), i.e. is lookup / insertion into `DictMap` of order `O(1)` or of `O(log n)`?

Reply:
`DictMap` is basically just a `IndexMap` [alias](https://docs.rs/rustworkx-core/latest/rustworkx_core/dictmap/type.DictMap.html) with a custom default hash algorithm. So lookup/insertion is `O(1)` but unlike a `HashMap` it retains insertion order. I went for using a dict map here because there is no guarantee that the node indices are contiguous or a sequence. The simplest example of that is if the input graph is a `StableGraph` with nodes removed in the middle. In general I'd expect a `Vec` would be faster here, but since we need this function to be general to handle all sorts of different types of graphs this seemed like the best option for the return type.

Reply:
It’s slower than a vector but still pretty fast (`O(1)`). It keeps the insertion order but that is what users expect from a Python dictionary-like structure